src/core/DY.Core.Trace.Base.fst

```diff
@@ -697,3 +748,102 @@ val trace_forget_labels:
 let trace_forget_labels tr =
   fmap_trace forget_label tr
 
+
+(*** Trace Searching ***)
+
+#push-options "--fuel 2 --ifuel 1"
+/// Finds the first (oldest) trace entry satisfying a given predicate.
+[@@"opaque_to_smt"]
+val trace_search_first: #label_t:Type -> tr:trace_ label_t -> p:(trace_entry_ label_t -> bool) ->
+    Pure (option timestamp)
+    (requires True)
+    (ensures fun ts_opt ->
+       match ts_opt with
+       | None -> forall ts. ts `on_trace` tr ==> ~(p(get_entry_at tr ts))
+       | Some ts -> (
+         ts `on_trace` tr /\
+         p (get_entry_at tr ts) /\
+         forall ts'. (ts' `on_trace` tr /\ p (get_entry_at tr ts')) ==> ts' >= ts
+       )
+    )
+let rec trace_search_first tr p
+  = match tr with
+    | Nil -> None
+    | Snoc hd entry ->
+      match trace_search_first hd p with
+      | None -> if p entry
+               then begin
+                 // This assertion seems to be necessary for proof performance
+                 // to avoid flakiness.
+                 assert(forall ts'. ts' `on_trace` hd ==> ~(p(get_entry_at tr ts')));
+                 Some (last_timestamp tr)
```

Comment:
This and the comparable point in `trace_search_last` seem to be the source of flakiness. I'm not entirely sure why --- both proved fine in a simpler form on my machine, so I assume it's nearing a resource limit, and this does appear to resolve it, but I would much rather figure out a way to not need this assertion (and, in the case of `trace_search_last`, to not need to match on the return value of the recursive call).

Reply:
Additionally, the assertion is almost just a part of the postcondition of the recursive call, just substituting `tr` for `hd` on the right-hand side of the implication (maybe this is triggering `get_entry_at_grows` in a way that wasn't happening before?)

Reply:
I notice that there is an rlimit of 25 on `find_event_triggered_at_timestamp_opt`

Reply:
I pushed a commit that fixes flakiness for good, the problem is in the `forall`: the type inferred by F* was probably a refined timestamp (because of the call with `get_event_at`), quantifying on `ts:timestamp` (without refinement) allows for better triggering.

Reply:
Looks good, thanks.